### PR TITLE
PALO ALTO NETWORKS threat vault

### PR DESCRIPTION
**Pull request by bakatzir**
## Status
- [x] Ready

## TODO:
with AF in same pack - or in a single pack? - **keeping in a different pack.**
rm `threatvault-antispyware-signature-get-by-id`, `threatvault-dns-signature-get-by-id`? **No, keep them.**

## Done:
Direct Searches
- [x] threatvault-antivirus-signtature-get (includes wildfire as well) - ips/signature/{signature_id}
sha256
https://docs.paloaltonetworks.com/autofocus/autofocus-api/perform-direct-searches/get-vulnerability-signature.html#idf286ab39-34f6-4f2a-9577-37f3120ad2b4

- [x] threatvault-dns-signature-get-by-id - dns/signature/{DNS_RTDNS_signature_id}
dns_signature_id
https://docs.paloaltonetworks.com/autofocus/autofocus-api/perform-direct-searches/get-anti-spyware-signature.html#idecf1edca-5929-4719-a924-8b1c6c1b541e

- [x] threatvault-antispyware-signature-get-by-id - ips/release/{release_id}
signature_id
https://docs.paloaltonetworks.com/autofocus/autofocus-api/perform-direct-searches/get-antivirus-signature.html#idf4e826c5-5e69-4e26-9016-86ea6cdd3cbd

- [x] threatvault-ip-geo-get
ip_address
https://docs.paloaltonetworks.com/autofocus/autofocus-api/perform-direct-searches/get-geolocation.html#idd6af5057-a0dd-42f5-9204-64aebf3e2954

- [x] add `!ip` with dbot score 0 and the location, ask regarding feedname and feedsource. consider implementing this: https://github.com/demisto/etc/issues/26418 - screenshot 104, 105

Search for Signatures 
- [x] threatvault-antivirus-signature-search /panav/search
signature_name
from
size

- [x] threatvault-dns-signature-search /dns/search
signature_name
domain_name
from
size

- [x] threatvault-antispyware-signature-search /ips/search
signature_name
vendor
cve
from
size

 - [x] threatvault-signature-search-result View Search Results
type antivirus/dns/antispyware (map to the relevant API - panav/dns/ips)
search_id

- [x] review all yml outputs
- [x] add Generic Polling playbook for conducting search  - in screenshots 101
- [x] add !file for the antivirus- in screenshots 102, 103
- [x] in `-search` cmds, rm page term, use from, to args. handle HR and outputs accordingly to reflect that.

### 101
Screen Shot 2020-11-03 at 11 57 23
### 102
Screen Shot 2020-11-04 at 12 10 11
### 103
Screen Shot 2020-11-04 at 12 08 17
### 104
Screen Shot 2020-11-05 at 12 37 03
### 105
Screen Shot 2020-11-05 at 12 38 01

## Related Issues
fixes: https://github.com/demisto/etc/issues/14540

## Minimum version of Demisto
- [x] 5.0.0

## Does it break backward compatibility?
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/Threat_Vault/Threat_Vault.py [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+
+
+def main():
+    """
+        PARSE AND VALIDATE INTEGRATION PARAMS
+    """
+    params = demisto.params()
+    api_key = params.get('api_key')
+    verify = not params.get('insecure', False)
+    proxy = params().get('proxy') is True
```
Comment (Itay4) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/Threat_Vault/Threat_Vault.py:455]:
```suggestion
    proxy = params.get('proxy')
```

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/Threat_Vault/Threat_Vault.py [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+            cve: cve ID.
+        Returns:
+            Response from API.
+        """
+        if path == 'dns':  # DNS search
+            if signature_name and domain_name:
+                raise Exception('Please provide either a signature_name or a domain_name.')
+        elif path == 'ips':  # Anti spyware search
+            if (cve and (vendor or signature_name)) or (vendor and (cve and signature_name)) \
+                    or (signature_name and (cve or vendor)):
```
Comment (yuvalbenshalom) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/Threat_Vault/Threat_Vault.py:90]:
```suggestion
            if (cve and (vendor or signature_name)) or (vendor and (cve or signature_name)) \
                    or (signature_name and (cve or vendor)):
```
maybe this is what you meant?
consider covering this in UT.

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/Threat_Vault/Threat_Vault.py [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+            search_request_id: signature id.
+        Returns:
+            Response from API.
+        """
+        return self._http_request(method='GET',
+                                  url_suffix=f'/threatvault/{search_type}/search/result/{search_request_id}',
+                                  params=self._params)
+
+
+def test_module(client: Client, *_) -> str:
```
Comment (yuvalbenshalom) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/Threat_Vault/Threat_Vault.py:127]: I think it is better to move it closer to main.

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/Threat_Vault/Threat_Vault.py [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+    """Performs basic get request to get a DNS signature.
+
+    Args:
+        client: Client object with request.
+
+    Returns:
+        string.
+    """
+    client.dns_signature_get_request(dns_signature_id='325235352')
```
Comment (yuvalbenshalom) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/Threat_Vault/Threat_Vault.py:136]: is this stable and will not get 404 in time?

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/Threat_Vault/Threat_Vault.yml [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+  dockerimage: demisto/python3:3.8.6.12176
+  isfetch: false
+  longRunning: false
+  longRunningPort: false
+  runonce: false
+  script: '-'
+  subtype: python3
+  type: python
+tests:
+- No test
```
Comment (yuvalbenshalom) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/Threat_Vault/Threat_Vault.yml:355]: why?

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/Threat_Vault/Threat_Vault.yml [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+fromversion: 5.0.0
```
Comment (yuvalbenshalom) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/Threat_Vault/Threat_Vault.yml:356]: need to check if we are should still release for 5.0.0 and not 5.5.0

**Review by yuvalbenshalom — CHANGES_REQUESTED** [review]
looks good. see my comments.

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/Threat_Vault/Threat_Vault.py (hunk 4 above)
Comment (bakatzir) [reply Packs/PaloAltoNetworks_Threat_Vault/Integrations/Threat_Vault/Threat_Vault.py:136]: I believe so, the signature ID should not change, will change it to an IP anyhow.

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/Threat_Vault/Threat_Vault.yml (hunk 6 above)
Comment (bakatzir) [reply Packs/PaloAltoNetworks_Threat_Vault/Integrations/Threat_Vault/Threat_Vault.yml:356]: I think we are still at 5.0.0 and above, will validate.

**Comment by ShirleyDenkberg** [comment]
@Itay4 Doc review is done.

**Review by yuvalbenshalom — APPROVED** [review]

**Pull request merged**